### PR TITLE
Move parameter validation specific to TPU Training plugins

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Follow up to #5441 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/core/decorators.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def parameter_validation(fn: Callable) -> Callable:
     when tying weights on TPU's.
 
     Args:
-        fn: ``.to`` method
+        fn: ``model_to_device`` method
```
Comment (awaelchli) [inline pytorch_lightning/core/decorators.py:79]: can you also update the docstring above that mentiones `LightningModule.to`?

**Review by awaelchli — APPROVED** [review]

**Comment by lezwon** [comment]
@kaushikb11 the TPU tests are being skipped. Probably the TPU device is not being detected :)

File: tests/accelerators/test_tpu_backend.py [hunk 2]
```diff
@@ -95,25 +95,21 @@ def test_weight_tying_warning(tmpdir, capsys=None):
...
+    Ensure no warning for parameter mismatch is thrown.
+    """
 
-#         def on_post_move_to_device(self):
-#             self.layer_3.weight = self.layer_1.weight
+    class Model(WeightSharingModule):
 
-#     model = Model()
-#     trainer = Trainer(checkpoint_callback=True, max_epochs=1, tpu_cores=1)
+        def on_post_move_to_device(self):
```
Comment (tchaton) [inline tests/accelerators/test_tpu_backend.py:108]: Could we make this check slightly smarter but checking parameters names ?

If I do self.layer_3.weight = self.layer_1.weight in the init function and mess up and do self.layer_3.weight = self.layer_2.weight, I won't get a warning but tying is different. Ideally it would be great to explicitly tell which weights are shared or do it automatically for the user.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 3]
```diff
@@ -171,6 +172,7 @@ def new_process(self, process_idx: int, trainer, mp_queue) -> None:
         if self.global_rank == 0:
             time.sleep(2)
 
+    @parameter_validation
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:175]: how sow is this?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/core/decorators.py [hunk 4]
```diff
@@ -71,12 +71,11 @@ def auto_transfer_args(self, *args, **kwargs):
 
 def parameter_validation(fn: Callable) -> Callable:
```
Comment (awaelchli) [inline pytorch_lightning/core/decorators.py:72]: I think now that you changed the decorator target to self.model, this decorator may no longer fit very well into core/decorators because it is basically now specific to the plugin having the attribute self.model. 
What do you think about moving it? 

Just for consideration

Comment (ananthsub) [reply pytorch_lightning/core/decorators.py:72]: +1 to @awaelchli 's suggestion

Comment (kaushikb11) [reply pytorch_lightning/core/decorators.py:72]: Good catch. Will do a follow-up PR for this.

File: tests/accelerators/test_tpu_backend.py (hunk 2 above)
Comment (kaushikb11) [reply tests/accelerators/test_tpu_backend.py:108]: Interesting, will follow up.
